src/Hopper/Internal/LoweredCore/ClosureConversion.hs

```diff
@@ -0,0 +1,281 @@
+--   'BinderInfo's.
+
+-- | Traversal focusing on the top 'EnvState' in our 'EnvStack'.
+topEnv :: Traversal' ConversionState EnvState
+topEnv = csEnvStack._head
+
+-- | Utility analogous to 'use' in lens. This utility should only be used when
+-- we're guaranteed to have a non-empty 'EnvStack'.
+topEnvUse :: Lens' EnvState a -> ConversionM a
+topEnvUse l = projectValue <$> get
+  where
+    projectValue state = fromMaybe (error "impossible env stack underrun") $
+      firstOf (topEnv.l) state
+
+-- | Pushes an empty 'EnvState' onto our 'EnvStack'.
+pushEmptyEnv :: ClosureType -> ConversionM ()
+pushEmptyEnv closureType = csEnvStack %= (emptyEnv:)
+  where
+    emptyEnv = EnvState mempty mempty (EnvSize 0) Map.empty closureType
+
+-- | Pops the top 'EnvState' off our our 'EnvStack'.
+popEnv :: ConversionM EnvState
+popEnv = do
+  envState <- topEnvUse id
+  csEnvStack %= tail
+  return envState
+
+-- | Issues a closure code ID to add the next closure to the registry.
+allocClosureCodeId :: ConversionM ClosureCodeId
+allocClosureCodeId = do
+  curr <- use csNextClosureId
+  csNextClosureId %= succ
+  return curr
+
+-- | Issues a thunk code ID to add the next thunk to the registry.
+allocThunkCodeId :: ConversionM ThunkCodeId
+allocThunkCodeId = do
+  curr <- use csNextThunkId
+  csNextThunkId %= succ
+  return curr
+
+-- TODO(bts): populate code records with real 'BinderInfo's
+dummyBI :: BinderInfo
+dummyBI = BinderInfoData Omega () Nothing
+
+-- | Adjusts a variable as a function of the number of 'Let's we've passed since
+-- entering the most immediate closure or thunk. We see whether the variable
+-- reaches outside of the current closure/thunk, and adjust the depth of the
+-- variable accordingly.
+adjustVar :: Word32 -> Variable -> ConversionM Variable
+adjustVar _letsPassed var@(GlobalVarSym _) = return var
+adjustVar letsPassed  var@(LocalVar lnv) = do
+  mClosureType <- firstOf (topEnv.esClosureType) <$> get
+
+  case mClosureType of
+    Nothing ->
+      return var
+    Just closureType ->
+      case reach closureType of
+        LocalReference -> return var
+        ArgReference -> return $ bump var
+        FreeReference depthBeyondClosure -> closeOver depthBeyondClosure
+
+  where
+    depth = lnv ^. lnDepth
+    slot = lnv ^. lnSlot
+
+    reach :: ClosureType -> Reach
+    reach Thunk
+      | depth >= letsPassed = FreeReference $ depth - letsPassed
+      | otherwise           = LocalReference
+    reach Closure
+      | depth >  letsPassed = FreeReference $ depth - (letsPassed + 1)
+      | depth == letsPassed = ArgReference
+      | otherwise           = LocalReference
+
+    bump :: Variable -> Variable
+    bump = localNameless.lnDepth %~ succ
+
+    closeOver :: Word32 -> ConversionM Variable
+    closeOver depthBeyondClosure = do
+      let envVar = LocalVar $ LocalNamelessVar depthBeyondClosure slot
+      mSharedEnvSlot <- topEnvUse $ esSlots.at envVar
+
```

Comment:
Ok. So would it also be correct to say "if a free variable is referred to
mutiple times and and are under the same lambda, they should be transformed
to point to the new closure slot variable"?

Or alternatively "every occurrence of a debrujin variable reference
pointing at the same free variable shall , by dint of being the same
variable , be transformed to point to the same environment/closure variable
slot"

The way I'm reading your description doesn't parse in my head as saying
those, hence my confusion.

On Thursday, April 21, 2016, Brian Schroeder <notifications@github.com>
wrote:

> In src/Hopper/Internal/LoweredCore/ClosureConversion.hs
> <https://github.com/hopper-lang/hopper/pull/14#discussion_r60586567>:
>
> > +    reach Thunk
> > +      | depth >= letsPassed = FreeReference $ depth - letsPassed
> > +      | otherwise           = LocalReference
> > +    reach Closure
> > +      | depth >  letsPassed = FreeReference $ depth - (letsPassed + 1)
> > +      | depth == letsPassed = ArgReference
> > +      | otherwise           = LocalReference
> > +
> > +    bump :: Variable -> Variable
> > +    bump = localNameless.lnDepth %~ succ
> > +
> > +    closeOver :: Word32 -> ConversionM Variable
> > +    closeOver depthBeyondClosure = do
> > +      let envVar = LocalVar $ LocalNamelessVar depthBeyondClosure slot
> > +      mSharedEnvSlot <- topEnvUse $ esSlots.at envVar
> > +
>
> i'm using "shared" here to mean already-allocated env slot that we can
> re-use because >1 variable refers to the same binder outside of the
> closure. see this test
> <https://github.com/hopper-lang/hopper/pull/14/files#diff-b0fda6d36a4df822c743158c87926130R230>.
> note that while the test re-uses an env slot for two variables that happen
> to be at the same depth and have the same debruijn index, if one of the
> variables was nested under another let but still refered to the same binder
> as the other variable, they would also share the same env slot.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/hopper-lang/hopper/pull/14/files/522546c3e91e583f2ff69a0e970b394917596bb4#r60586567>
>